modMcf/test/org/aion/mcf/db/IContractDetailsTest.java

```diff
@@ -0,0 +1,417 @@
+
+    @Before
+    public void setup() {
+        IContractDetails<IDataWord> details = new AionContractDetailsImpl();
+        cache1 = new ContractDetailsCacheImpl(details);
+        details = new ContractDetailsCacheImpl(details);
+        cache2 = new ContractDetailsCacheImpl(details);
+    }
+
+    @After
+    public void tearDown() {
+        cache1 = null;
+        cache2 = null;
+    }
+
+    @Test
+    public void testGetNoSuchSingleKey() {
+        doGetNoSuchSingleKeyTest(cache1);
+        doGetNoSuchSingleKeyTest(cache2);
+    }
+
+    @Test
+    public void testGetNoSuchDoubleKey() {
+        doGetNoSuchDoubleKeyTest(cache1);
+        doGetNoSuchDoubleKeyTest(cache2);
+    }
+
+    @Test
+    public void testPutSingleZeroValue() {
+        IDataWord key = new DataWord(RandomUtils.nextBytes(DataWord.BYTES));
+        doPutSingleZeroValueTest(cache1, key);
+        doPutSingleZeroValueTest(cache2, key);
+
+        key = new DoubleDataWord(RandomUtils.nextBytes(DoubleDataWord.BYTES));
+        doPutSingleZeroValueTest(cache1, key);
+        doPutSingleZeroValueTest(cache2, key);
+    }
+
+    @Test
+    public void testPutDoubleZeroValue() {
+        IDataWord key = new DataWord(RandomUtils.nextBytes(DataWord.BYTES));
+        doPutDoubleZeroValueTest(cache1, key);
+        doPutDoubleZeroValueTest(cache2, key);
+
+        key = new DoubleDataWord(RandomUtils.nextBytes(DoubleDataWord.BYTES));
+        doPutDoubleZeroValueTest(cache1, key);
+        doPutDoubleZeroValueTest(cache2, key);
+    }
+
+    @Test
+    public void testPutSingleZeroKey() {
+        IDataWord value = new DataWord(RandomUtils.nextBytes(DataWord.BYTES));
+        doPutSingleZeroKeyTest(cache1, value);
+        doPutSingleZeroKeyTest(cache2, value);
+
+        value = new DoubleDataWord(RandomUtils.nextBytes(DoubleDataWord.BYTES));
+        doPutSingleZeroKeyTest(cache1, value);
+        doPutSingleZeroKeyTest(cache2, value);
+    }
+
+    @Test
+    public void testPutDoubleZeroKey() {
+        IDataWord value = new DataWord(RandomUtils.nextBytes(DataWord.BYTES));
+        doPutDoubleZeroKeyTest(cache1, value);
+        doPutDoubleZeroKeyTest(cache2, value);
+
+        value = new DoubleDataWord(RandomUtils.nextBytes(DoubleDataWord.BYTES));
+        doPutDoubleZeroKeyTest(cache1, value);
+        doPutDoubleZeroKeyTest(cache2, value);
+    }
+
+    @Test
+    public void testPutZeroKeyAndValue() {
+        // Try single-single
+        cache1.put(DataWord.ZERO, DataWord.ZERO);
+        assertNull(cache1.get(DataWord.ZERO));
+        cache2.put(DataWord.ZERO, DataWord.ZERO);
+        assertNull(cache2.get(DataWord.ZERO));
+
+        // Try single-double
+        cache1.put(DataWord.ZERO, DoubleDataWord.ZERO);
+        assertNull(cache1.get(DataWord.ZERO));
+        cache2.put(DataWord.ZERO, DoubleDataWord.ZERO);
+        assertNull(cache2.get(DataWord.ZERO));
+
+        // Try double-single
+        cache1.put(DoubleDataWord.ZERO, DataWord.ZERO);
```

Comment:
Thanks for the detailed test cases -- I'm questioning a bit about whether it's necessary to try all the permutations of double/single though.  The put/get methods basically just put/get from a Map<IDataWord, IDataWord> and the code path doesn't really seem to care at all about what specific kind of IDataWord is in the key or value.  Do you think all the combinations need to be tested?

Also, I noticed a few other public methods that don't seem to be tseted at the moment, i.e. ContractDetailsCacheImpl#commit() and some other ones.  Do those not need to be tested?  Or did we just not have time to get to them yet?

Reply:
I like to have the permutations tested just so that if implementation details ever change we can still be confident that this is all working correctly.

My priority for these tests was just to fix how the zero-bytes input was handled when adding it to the db and querying it, so I left out tests unrelated to this atm. But you're right about commit, I missed that one, I'll add some tests for that